reframe/frontend/cli.py

```diff
@@ -111,6 +111,124 @@ def generate_report_filename(filepatt):
+
+    # Filter checks by prgenv
+    if not options.skip_prgenv_check:
+        for prgenv in options.prgenv:
+            checks_matched = filter(filters.have_prgenv(prgenv),
+                                    checks_matched)
+
+    # Filter checks by system
+    if not options.skip_system_check:
+        checks_matched = filter(
+            filters.have_partition(rt.system.partitions), checks_matched)
+
+    if options.gpu_only:
+        checks_matched = filter(filters.have_gpu_only(), checks_matched)
+    elif options.cpu_only:
+        checks_matched = filter(filters.have_cpu_only(), checks_matched)
+
+    # Determine the allowed programming environments
+    allowed_environs = {e.name
+                        for env_patt in options.prgenv
+                        for p in rt.system.partitions
+                        for e in p.environs if re.match(env_patt, e.name)}
+
+    # Generate the test cases, validate dependency and sort them
+    checks_matched = list(checks_matched)
+
+    return checks_matched
+
+# TODO place this function in a proper module
+# TODO define mechanism on how to propagate command line options, currently
+#      we are passing site_config, but that's clearly a bad design
+#      read below the options we need to propagate or make available to each
+#      ci invokation
+#      An alternative design would be to have environment variables exported
+#      during the trigger phase of reframe that would influence the tests
+#      individually
+def generate_ci_pipeline_file(ci_pipeline_file, ci_tags, checks_found, options, rt, loader, site_config):
+    import reframe.utility as util
+    import yaml
+
+    stages = {
+        'stages': [f'rfm-stage-{m}' for m in ci_tags]
+    }
+
+    with open(ci_pipeline_file, 'w') as pipeline_file:
+        for entry in yaml.safe_dump(stages, indent=2).split('\n'):
+            pipeline_file.write(f'{entry}\n')
+
+    for tag in ci_tags:
+        pipeline = {}
+        checks_matched = filter_checks(checks_found, options, rt, loader, ci_tag=tag)
+
+        # TODO there are some options that need  to be revisited
+        # - about the test folder: reframe's -c option
+        # - about the configuration file to use: reframe's -C option
+        # - about the artifacts artifacts: {paths: [jobs_scratch_dir], when: always}
+
+        # TODO there are some missing options:
+        # - about other potentially important options:
+        # + --keep-stage-files
+        # + --ignore-check-conflicts
+        # + -p
+        # + --gpu-only
+        # + --cpu-only
+        # + -A
+        # + -P
+        # + --reservation
+        # + --skip-performance-check
+        # + --nodelist
+        # + --exclude-nodes
+        # + --skip-system-check
+        # + --skip-prgenv-check
+        # + --mode
+        # + --max-retries
+        # + -M
+        # + -m
+        # + --module-mappings
+        # + --failure-stats
+        # + --performance-report
+        for test in checks_matched:
```

Comment:
Everything must happen at the test case level. Now the test cases are generated immediately and everything in the frontend operates on them.